exp/api/remote/remote_api.go

```diff
@@ -0,0 +1,530 @@
+	return func(o *apiOpts) error {
+		o.path = path
+		return nil
+	}
+}
+
+// WithAPIRetryOnRateLimit returns APIOption that disables retrying on rate limit status code.
+func WithAPINoRetryOnRateLimit() APIOption {
+	return func(o *apiOpts) error {
+		o.retryOnRateLimit = false
+		return nil
+	}
+}
+
+// WithAPIBackoff returns APIOption that allows overriding backoff configuration.
+func WithAPIBackoff(backoff backoff.Config) APIOption {
+	return func(o *apiOpts) error {
+		o.backoff = backoff
+		return nil
+	}
+}
+
+type nopSlogHandler struct{}
+
+func (n nopSlogHandler) Enabled(context.Context, slog.Level) bool  { return false }
+func (n nopSlogHandler) Handle(context.Context, slog.Record) error { return nil }
+func (n nopSlogHandler) WithAttrs([]slog.Attr) slog.Handler        { return n }
+func (n nopSlogHandler) WithGroup(string) slog.Handler             { return n }
+
+// NewAPI returns a new API for the clients of Remote Write Protocol.
+//
+// It is not safe to use the returned API from multiple goroutines, create a
+// separate *API for each goroutine.
+func NewAPI(client *http.Client, baseURL string, opts ...APIOption) (*API, error) {
+	parsedURL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid base URL: %w", err)
+	}
+
+	o := *defaultAPIOpts
+	for _, opt := range opts {
+		if err := opt(&o); err != nil {
+			return nil, err
+		}
+	}
+
+	if o.logger == nil {
+		o.logger = slog.New(nopSlogHandler{})
+	}
+
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	parsedURL.Path = path.Join(parsedURL.Path, o.path)
+
+	return &API{
+		opts:    o,
+		client:  client,
+		baseURL: parsedURL,
+	}, nil
+}
+
+type retryableError struct {
+	error
+	retryAfter time.Duration
+}
+
+func (r retryableError) RetryAfter() time.Duration {
+	return r.retryAfter
+}
+
+type vtProtoEnabled interface {
+	SizeVT() int
+	MarshalToSizedBufferVT(dAtA []byte) (int, error)
+}
+
+type gogoProtoEnabled interface {
+	Size() (n int)
+	MarshalToSizedBuffer(dAtA []byte) (n int, err error)
+}
+
+// Sort of a hack to identify v2 requests.
+// Under any marshaling scheme, v2 requests have a `Symbols` field of type []string.
+// So would always have a `GetSymbols()` method which doesn't rely on any other types.
+type v2Request interface {
+	GetSymbols() []string
+}
+
+// Write writes given, non-empty, protobuf message to a remote storage.
+//
+// Depending on serialization methods,
+//   - https://github.com/planetscale/vtprotobuf methods will be used if your msg
+//     supports those (e.g. SizeVT() and MarshalToSizedBufferVT(...)), for efficiency
+//   - Otherwise https://github.com/gogo/protobuf methods (e.g. Size() and MarshalToSizedBuffer(...))
+//     will be used
+//   - If neither is supported, it will marshaled using generic google.golang.org/protobuf methods and
+//     error out on unknown scheme.
+func (r *API) Write(ctx context.Context, msg any) (_ WriteResponseStats, err error) {
+	// Reset the buffer.
+	r.reqBuf = r.reqBuf[:0]
```

Comment:
This is good for single thread, but people might want to use the same http client etc for multiple concurrent streams, so we probably can't do this. I would remove and in future add (without breaking variadic options for write, we could do it now too it's a common with `WithWriteBuffer`)